gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java

```diff
@@ -0,0 +1,322 @@
+          + COL_GC_RUN_START
+          + ")";
+
+  private final SparkSession sparkSession;
+  private final String catalogAndTable;
+  private final String gcRefName;
+  private final StructType schema;
+
+  public IdentifiedResultsRepo(
+      SparkSession sparkSession, String catalog, String gcRefName, String gcTable) {
+    this.sparkSession = sparkSession;
+    this.catalogAndTable = String.format("%s.%s", catalog, gcTable);
+    this.gcRefName = gcRefName;
+    this.schema =
+        checkAndCreateTable(
+            sparkSession, catalog, TableIdentifier.parse(gcTable), catalogAndTable, gcRefName);
+  }
+
+  public StructType getSchema() {
+    return schema;
+  }
+
+  /**
+   * Collect the expired contents for the given run id as {@link IdentifiedResult} .
+   *
+   * @param runId run id of completed identify task.
+   * @return {@link IdentifiedResult} object having the expired contents per content id per
+   *     reference.
+   */
+  public IdentifiedResult collectExpiredContents(String runId) {
+    List<Row> rows = collectExpiredContentsAsDataSet(runId).collectAsList();
+    IdentifiedResult identifiedResult = new IdentifiedResult();
+    ObjectMapper objectMapper = new ObjectMapper();
+    rows.forEach(row -> addRowToIdentifiedResults(identifiedResult, objectMapper, row));
+    return identifiedResult;
+  }
+
+  /**
+   * Collect the expired contents for the given run id as spark dataset.
+   *
+   * @param runId run id of completed identify task.
+   * @return spark dataset of row where each row is having the expired contents per content id per
+   *     reference.
+   */
+  public Dataset<Row> collectExpiredContentsAsDataSet(String runId) {
+    // collect marker row for the given run-id
+    String checkMarkerRow =
+        String.format(
+            "SELECT %s FROM %s WHERE %s = '%s' AND %s = '%s'",
+            COL_GC_RUN_ID,
+            withRefName(catalogAndTable, gcRefName),
+            COL_GC_RUN_ID,
+            runId,
+            COL_TYPE,
+            TYPE_GC_MARKER);
+
+    List<Row> markerRows = sparkSession.sql(checkMarkerRow).collectAsList();
+    if (markerRows.isEmpty()) {
+      return sparkSession.createDataFrame(Collections.emptyList(), schema);
+    }
+
+    String expiredContentIdsQuery =
+        String.format(
+            "SELECT %s, %s, %s FROM %s WHERE %s = '%s' AND %s = '%s'",
+            COL_CONTENT_ID,
+            COL_EXPIRED_CONTENTS,
+            COL_EXPIRED_AT_REFERENCE_NAME,
+            //
+            withRefName(catalogAndTable, gcRefName),
+            //
+            COL_GC_RUN_ID,
+            runId,
+            //
+            COL_TYPE,
+            TYPE_CONTENT);
+
+    return sparkSession.sql(expiredContentIdsQuery);
+  }
+
+  /**
+   * Get the last written run id which is having a marker row.
+   *
+   * @return latest completed run id.
+   */
+  public String getLatestCompletedRunID() {
+    // collect marker row for the last written run id
+    String checkMarkerRow =
```

Comment:
why not do a groupby and max? Prob faster and better on memory than a sort

Reply:
If there is an in-progress entry (without marker row), max(timestamp) will select the timestamp of in-progress entry and I may not find marker row for that timestamp and query will return empty results.

So, I think current SQL is safe.